File: app/src/main/java/com/zibete/proyecto1/ui/ChatList/ChatListFragment.java
package com.zibete.proyecto1.ui.ChatList;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.airbnb.lottie.LottieAnimationView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.zibete.proyecto1.adapters.AdapterChatLista;
import com.zibete.proyecto1.Constants;
import com.zibete.proyecto1.model.ChatWith;
import com.zibete.proyecto1.R;
import com.zibete.proyecto1.utils.UserRepository;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;

import static com.zibete.proyecto1.Constants.Empty;
import static com.zibete.proyecto1.Constants.FRAGMENT_ID_CHATGROUPLIST;
import static com.zibete.proyecto1.Constants.FRAGMENT_ID_CHATLIST;
import static com.zibete.proyecto1.Constants.chatWith;
import static com.zibete.proyecto1.Constants.chatWithUnknown;
import static com.zibete.proyecto1.utils.FirebaseRefs.refDatos;
import static com.zibete.proyecto1.ui.ChatList.ChatListGroupsFragment.chatsGroupArrayList;

public class ChatListFragment extends Fragment implements SearchView.OnQueryTextListener{

    LottieAnimationView lottieChatRight, lottieChatLeft;
    ProgressBar progressbar;
    RecyclerView rv;
    ArrayList <ChatWith> chatsArrayList = new ArrayList<>();
    LinearLayout linearOnBoardingChatList;
    LinearLayout tv_linearOnBoardingChatList;

    ArrayList<ChatWith> getSorted(ArrayList<ChatWith> chatsArrayList2) {

        ArrayList<ChatWith> list = chatsArrayList2;

        Collections.sort(list, new Comparator<ChatWith>() {
            @Override
            public int compare(ChatWith o1, ChatWith o2) {
                return o1.getDate().compareTo(o2.getDate());
            }
        });

        return list;
    }


    private AdapterChatLista adapterChatLista;
    LinearLayoutManager mLayoutManager;
    FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();

    public ChatListFragment() {
        //Constructor vacío
    }


    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_chat_list, container, false);

        setHasOptionsMenu(true);

        progressbar = view.findViewById(R.id.progressbar2);
        linearOnBoardingChatList = view.findViewById(R.id.linearOnBoardingChatList);
        mLayoutManager = new LinearLayoutManager(getContext());

        mLayoutManager.setReverseLayout(true);
        mLayoutManager.setStackFromEnd(true);

        rv = view.findViewById(R.id.rv);
        rv.setLayoutManager(mLayoutManager);
        adapterChatLista = new AdapterChatLista(chatsArrayList, getContext());
        rv.setAdapter(adapterChatLista);
        registerForContextMenu(rv);

        lottieChatRight = view.findViewById(R.id.lottieChatRight);
        lottieChatLeft = view.findViewById(R.id.lottieChatLeft);




/*
        DisplayMetrics metrics = new DisplayMetrics();
        WindowManager windowManager = (WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE);
        windowManager.getDefaultDisplay().getMetrics(metrics);

        int widthPixels = metrics.widthPixels; //Ancho
        int heightPixels = metrics.heightPixels; //Alto

        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT, widthPixels/2);

        linearOnBoardingChatList.setLayoutParams(layoutParams);

 */











//CHAT LISTA
//CREAR LISTA

        refDatos.child(user.getUid()).child(chatWith).addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String previousChildName) {


                if (dataSnapshot.exists()) {


                        ChatWith chat = dataSnapshot.getValue(ChatWith.class);
                        adapterChatLista.addChats(chat);



                    @SuppressLint("SimpleDateFormat") final SimpleDateFormat dateFormat3 = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
                    for (ChatWith wChat : chatsArrayList) {


                        try {
                            Date date = dateFormat3.parse(wChat.getDateTime());
                            wChat.setDate(date);
                        } catch (ParseException e) {
                            e.printStackTrace();
                        }

                    }
                }
                Collections.sort(chatsArrayList);


            }

            //CADA VEZ QUE CAMBIA ALGO
            @Override
            public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String previousChildName) {


                ValueEventListener listenerChangeChatList = new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                        if (dataSnapshot.exists()) {

                            ArrayList <ChatWith> chatsArrayList2 = new ArrayList<>();
                            chatsArrayList2.clear();


                            for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                                ChatWith chat = snapshot.getValue(ChatWith.class);
                                chatsArrayList2.add(chat);

                            }

                            @SuppressLint("SimpleDateFormat") final SimpleDateFormat dateFormat3 = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
                            for (ChatWith wChat : chatsArrayList2){

                                try {
                                    Date date = dateFormat3.parse(wChat.getDateTime());
                                    wChat.setDate(date);
                                } catch (ParseException e) {
                                    e.printStackTrace();
                                }

                            }

                            Collections.sort(chatsArrayList2);

                            adapterChatLista.updateData(chatsArrayList2);

                        } else {

                            progressbar.setVisibility(View.GONE);

                        }
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {

                    }
                };

                refDatos.child(user.getUid()).child(chatWith).addValueEventListener(listenerChangeChatList);

            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {


                ChatWith chatWith = dataSnapshot.getValue(ChatWith.class);
                adapterChatLista.deleteChat(chatWith);
            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot snapshot, @Nullable String previousChildName) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });






/*
        ref_datos.child(user.getUid()).child("ChatList").child("nuevoMensaje").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    final Integer newChats = dataSnapshot.getValue(Integer.class);
                    if (newChats > 0) {
                        dataSnapshot.getRef().setValue(0);
                    }
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
            }
        });

 */




        adapterChatLista.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                super.onItemRangeInserted(positionStart, itemCount);
                setScrollbar();
                progressbar.setVisibility(View.GONE);
            }
        });





        //Mostrar NO HAS CHATEADO
        refDatos.child(user.getUid()).child(chatWith).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {

                    long count = 0;

                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {

                        String state = snapshot.child("estado").getValue(String.class);
                        String photo = snapshot.child("wUserPhoto").getValue(String.class);

                        if (!photo.equals(Empty)) {

                            if (state.equals(chatWith) || state.equals("silent")) {

                                count = count + 1;

                            }
                        }
                    }

                    if (count == 0) {
                        rv.setVisibility(View.GONE);
                        linearOnBoardingChatList.setVisibility(View.VISIBLE);
                        lottieChatLeft.playAnimation();
                        new Handler().postDelayed(new Runnable(){
                            public void run(){
                                lottieChatRight.playAnimation();
                            }
                        }, 100);

                    }else{
                        rv.setVisibility(View.VISIBLE);
                        linearOnBoardingChatList.setVisibility(View.GONE);
                        lottieChatLeft.cancelAnimation();
                        lottieChatRight.cancelAnimation();
                    }
                }else{
                    rv.setVisibility(View.GONE);
                    linearOnBoardingChatList.setVisibility(View.VISIBLE);
                    lottieChatLeft.playAnimation();
                    new Handler().postDelayed(new Runnable(){
                        public void run(){
                            lottieChatRight.playAnimation();
                        }
                    }, 100);
                }
                progressbar.setVisibility(View.GONE);

            }
            @Override public void onCancelled(@NonNull DatabaseError error) {
            }
        });



        rv.setVisibility(View.VISIBLE);


        return view;
    }


    private void setScrollbar(){
        rv.scrollToPosition(adapterChatLista.getItemCount()-1);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {


        if(item.getGroupId() == FRAGMENT_ID_CHATLIST) {

            String type = chatWith;
            final ChatWith wChat = chatsArrayList.get(item.getOrder());
            String id_user = wChat.getUserId();
            String name_user = wChat.getUserName();
            RunItemSelected(item, type, id_user, name_user);
        }

        if(item.getGroupId() == FRAGMENT_ID_CHATGROUPLIST) {
            final ChatWith wChat = chatsGroupArrayList.get(item.getOrder());
            String id_user = wChat.getUserId();
            String name_user = wChat.getUserName();
            String type = chatWithUnknown;
            RunItemSelected(item, type, id_user, name_user);
        }



        return true;
    }

    public void RunItemSelected(MenuItem item, String type, String id_user, String name_user) {

        View view = getActivity().findViewById(android.R.id.content);
        switch (item.getItemId()) {
            case 1: //Marcar leído, no leído

                UserRepository.setNoLeido(id_user,type);
                break;

            case 2: //Silenciar notificaciones

                UserRepository.Silent(name_user, id_user, type);

                break;

            case 3: //Bloquear


                UserRepository.setBlockUser(getContext(), name_user, id_user, view, type);

                break;


            case 4: //Ocultar

                new Constants().UnhiddenChat(getContext(), id_user, name_user, view, type);

                break;


            case 5: //Eliminar

                new Constants().DeleteChat(getContext(), id_user, name_user, view, type);

                break;
        }
    }


    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        adapterChatLista.getFilter().filter(newText);
        return false;
    }

/*
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.main, menu);

        super.onCreateOptionsMenu(menu, inflater);
    }

 */

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);

        final MenuItem action_search = menu.findItem(R.id.action_search);
        final MenuItem action_settings2 = menu.findItem(R.id.action_unlock);
        final MenuItem action_exit = menu.findItem(R.id.action_exit);

        action_exit.setVisible(false);
        action_search.setVisible(true);
        action_settings2.setVisible(true);

        SearchView searchView = (SearchView) action_search.getActionView();

        searchView.setOnQueryTextListener(this);


    }




}

File: app/src/main/java/com/zibete/proyecto1/ui/ChatListFragment.kt
package com.zibete.proyecto1.ui

import android.R
import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.zibete.proyecto1.Constants
import com.zibete.proyecto1.adapters.AdapterChatLista
import com.zibete.proyecto1.adapters.ChatListGroupsFragment
import com.zibete.proyecto1.databinding.FragmentChatListBinding
import com.zibete.proyecto1.model.ChatWith
import com.zibete.proyecto1.utils.FirebaseRefs
import com.zibete.proyecto1.utils.UserRepository
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Collections

class ChatListFragment : Fragment(), SearchView.OnQueryTextListener {

    private var _binding: FragmentChatListBinding? = null
    private val binding get() = _binding!!

    private lateinit var adapterChatLista: AdapterChatLista
    private lateinit var layoutManager: LinearLayoutManager

    private val chatsArrayList: ArrayList<ChatWith> = ArrayList()

    // Listeners para limpiarlos correctamente
    private var chatListChildListener: ChildEventListener? = null
    private var emptyStateListener: ValueEventListener? = null

    private val currentUser
        get() = FirebaseAuth.getInstance().currentUser

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentChatListBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)

        val u = currentUser
        if (u == null) {
            setupInitialUi()
            showOnBoarding()
            return binding.root
        }

        setupRecycler()
        setupInitialUi()
        setupChatListListener(u.uid)
        setupEmptyStateListener(u.uid)
        setupAdapterObserver()

        return binding.root
    }

    // ---------- UI inicial ----------

    private fun setupInitialUi() = with(binding) {
        rv.isVisible = true
        linearOnBoardingChatList.isVisible = false
        progressbar2.isVisible = true

        lottieChatLeft.cancelAnimation()
        lottieChatRight.cancelAnimation()
    }

    private fun setupRecycler() {
        layoutManager = LinearLayoutManager(requireContext()).apply {
            reverseLayout = true
            stackFromEnd = true
        }

        adapterChatLista = AdapterChatLista(chatsArrayList, requireContext())

        binding.rv.apply {
            layoutManager = this@ChatListFragment.layoutManager
            adapter = adapterChatLista
        }

        registerForContextMenu(binding.rv)
    }

    // ---------- Listener principal de lista de chats ----------

    private fun setupChatListListener(userId: String) {
        val ref = FirebaseRefs.refDatos.child(userId).child(Constants.chatWith)

        val listener = object : ChildEventListener {
            // Nuevo chat
            override fun onChildAdded(snapshot: DataSnapshot, previousChildName: String?) {
                if (!snapshot.exists()) return

                val chat = snapshot.getValue(ChatWith::class.java) ?: return
                adapterChatLista.addChats(chat)
                updateDatesAndSort(chatsArrayList)
            }

            // Cambios en chats
            override fun onChildChanged(snapshot: DataSnapshot, previousChildName: String?) {
                ref.addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        if (!dataSnapshot.exists()) {
                            _binding?.progressbar2?.isVisible = false
                            return
                        }

                        val updatedList = ArrayList<ChatWith>()
                        for (child in dataSnapshot.children) {
                            val chat = child.getValue(ChatWith::class.java)
                            if (chat != null) updatedList.add(chat)
                        }

                        updateDatesAndSort(updatedList)
                        adapterChatLista.updateData(updatedList)
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
            }

            // Chat eliminado
            override fun onChildRemoved(snapshot: DataSnapshot) {
                val chat = snapshot.getValue(ChatWith::class.java) ?: return
                adapterChatLista.deleteChat(chat)
            }

            override fun onChildMoved(snapshot: DataSnapshot, previousChildName: String?) {}
            override fun onCancelled(error: DatabaseError) {}
        }

        chatListChildListener = listener
        ref.addChildEventListener(listener)
    }

    @SuppressLint("SimpleDateFormat")
    private fun updateDatesAndSort(list: MutableList<ChatWith>) {
        val format = SimpleDateFormat("dd/MM/yyyy HH:mm:ss")
        list.forEach { chat ->
            try {
                chat.date = format.parse(chat.dateTime)
            } catch (_: ParseException) {
            }
        }
        Collections.sort(list)
    }

    // ---------- Empty state / onboarding ----------

    private fun setupEmptyStateListener(userId: String) {
        val ref = FirebaseRefs.refDatos.child(userId).child(Constants.chatWith)

        val listener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val b = _binding ?: return

                if (dataSnapshot.exists()) {
                    var count = 0L

                    for (snapshot in dataSnapshot.children) {
                        val state = snapshot.child("estado").getValue(String::class.java)
                        val photo = snapshot.child("wUserPhoto").getValue(String::class.java)

                        if (photo != Constants.Empty &&
                            (state == Constants.chatWith || state == "silent")
                        ) {
                            count++
                        }
                    }

                    if (count == 0L) {
                        showOnBoarding()
                    } else {
                        showChatList()
                    }
                } else {
                    showOnBoarding()
                }

                b.progressbar2.isVisible = false
            }

            override fun onCancelled(error: DatabaseError) {}
        }

        emptyStateListener = listener
        ref.addValueEventListener(listener)
    }

    private fun showOnBoarding() = with(binding) {
        rv.isVisible = false
        linearOnBoardingChatList.isVisible = true

        lottieChatLeft.playAnimation()
        Handler(Looper.getMainLooper()).postDelayed({
            lottieChatRight.playAnimation()
        }, 100)
    }

    private fun showChatList() = with(binding) {
        rv.isVisible = true
        linearOnBoardingChatList.isVisible = false

        lottieChatLeft.cancelAnimation()
        lottieChatRight.cancelAnimation()
    }

    // ---------- Observer para scroll y progress ----------

    private fun setupAdapterObserver() {
        adapterChatLista.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                super.onItemRangeInserted(positionStart, itemCount)
                setScrollbar()
                _binding?.progressbar2?.isVisible = false
            }
        })
    }

    private fun setScrollbar() {
        val b = _binding ?: return
        if (::adapterChatLista.isInitialized && adapterChatLista.itemCount > 0) {
            b.rv.scrollToPosition(adapterChatLista.itemCount - 1)
        }
    }

    // ---------- Context menu acciones ----------

    override fun onContextItemSelected(item: MenuItem): Boolean {
        // Chat individual
        if (item.groupId == Constants.FRAGMENT_ID_CHATLIST) {
            val type = Constants.chatWith
            val wChat = chatsArrayList[item.order]
            runItemSelected(item, type, wChat.userId, wChat.userName)
        }

        // Chat unknown / grupos
        if (item.groupId == Constants.FRAGMENT_ID_CHATGROUPLIST) {
            val wChat = ChatListGroupsFragment.chatsGroupArrayList[item.order]
            val type = Constants.chatWithUnknown
            runItemSelected(item, type, wChat.userId, wChat.userName)
        }

        return true
    }

    private fun runItemSelected(
        item: MenuItem,
        type: String,
        idUser: String,
        nameUser: String?
    ) {
        val view = requireActivity().findViewById<View>(R.id.content)

        when (item.itemId) {
            1 -> UserRepository.setNoLeido(idUser, type)
            2 -> UserRepository.Silent(nameUser, idUser, type)
            3 -> UserRepository.setBlockUser(requireContext(), nameUser, idUser, view, type)
            4 -> Constants().UnhiddenChat(requireContext(), idUser, nameUser, view, type)
            5 -> Constants().DeleteChat(requireContext(), idUser, nameUser, view, type)
        }
    }

    // ---------- SearchView ----------

    @Deprecated("Deprecated in Java")
    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)

        val actionSearch = menu.findItem(com.zibete.proyecto1.R.id.action_search)
        val actionUnlock = menu.findItem(com.zibete.proyecto1.R.id.action_unlock)
        val actionExit = menu.findItem(com.zibete.proyecto1.R.id.action_exit)

        actionExit.isVisible = false
        actionSearch.isVisible = true
        actionUnlock.isVisible = true

        val searchView = actionSearch.actionView as? SearchView
        searchView?.setOnQueryTextListener(this)
    }

    override fun onQueryTextSubmit(query: String?): Boolean = false

    override fun onQueryTextChange(newText: String?): Boolean {
        adapterChatLista.filter.filter(newText)
        return true
    }

    // ---------- Ciclo de vida ----------

    override fun onDestroyView() {
        val u = currentUser
        if (u != null) {
            val ref = FirebaseRefs.refDatos.child(u.uid).child(Constants.chatWith)
            chatListChildListener?.let { ref.removeEventListener(it) }
            emptyStateListener?.let { ref.removeEventListener(it) }
        }

        chatListChildListener = null
        emptyStateListener = null
        _binding = null
        super.onDestroyView()
    }
}
